packages/flutter/lib/src/widgets/animated_child_switcher.dart

```diff
@@ -125,31 +205,36 @@ class _AnimatedChildSwitcherState extends State<AnimatedChildSwitcher> with Tick
     super.dispose();
   }
 
+  bool get hasNewChild => widget.child != null;
+  bool get hasOldChild => _currentChild?.widget != null;
+
   @override
-  Widget build(BuildContext context) {
-    if (widget.child != _currentChild.widget) {
-      addEntry(true);
+  void didUpdateWidget(AnimatedChildSwitcher oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    if (hasNewChild != hasOldChild || hasNewChild && !Widget.canUpdate(widget.child, _currentChild.widget)) {
+      addEntry(animate: true);
     }
-    final List<Widget> children = <Widget>[];
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    final List<AnimatedChildSwitcherLayoutData> children = <AnimatedChildSwitcherLayoutData>[];
     for (_AnimatedChildSwitcherChildEntry child in _children) {
       children.add(
-        new FadeTransition(
-          opacity: child.animation,
-          child: child.widget,
+        new AnimatedChildSwitcherLayoutData(
+          widget.transitionBuilder(child.widget, child.animation),
+          child.animation,
         ),
       );
     }
     if (_currentChild != null) {
       children.add(
-        new FadeTransition(
-          opacity: _currentChild.animation,
-          child: _currentChild.widget,
+        new AnimatedChildSwitcherLayoutData(
+          widget.transitionBuilder(_currentChild.widget, _currentChild.animation),
+          _currentChild.animation,
         ),
       );
```

Comment:
something else we need to do here is make sure we put keys on each of these children, otherwise the animations will get all messed up -- e.g. if you have animations A, B, and C running, then A ends and you take it out, without keys, B will be synced with A, C will be synced with B, and all of them will lose the state of any descendants.

Reply:
OK, but how can I do that if the transition builder is the one creating the children?  I can add it in the default, but not enforce it.

Reply:
Should I use "KeyedSubtree" to do this?

Reply:
Yeah, KeyedSubtree. (Or pass the key to the child and require the contract to be that they must use the key, which is what AnimatedCrossFade does iirc, though that makes the API less nice.)